src/RFC-0141_Sparse_Merkle_Trees.md

```diff
@@ -0,0 +1,669 @@
+for every million hashes in a bitmap to indicate which hashes have  been deleted (3 MB) for a total of 99MB.
+
+This only gets worse with time. Over a period of a year, a busy blockchain might have 100,000,000 spent transaction 
+outputs. However, the UTXO set will grow far more slowly, and perhaps only 10x in size to 10 million outputs.
+
+The SMT tree requires serialising 640MB of data to recreate the root, whereas the MMMR now requires 3.6GB of data.
+
+## Implementation
+
+The proposed implementation assumes that a key-value store for the data exists. The Merkle tree is only 
+concerned with the index and the value hash, as opposed to the value itself.
+
+When constructing a new tree, a hashing algorithm is specified. As 
+indicated, the "values" provided to the tree must already be a hash, and should have been generated from a different 
+hashing algorithm to the one driving the tree, in order to prevent second pre-image attacks.
+
+To insert a new leaf, the key is used to derive a path through the tree. Starting with the most significant bit, you 
+move down the left branch if the bit is zero, or take the right branch if the bit is equal to one. Once a terminal 
+node is reached, the node is replaced with a new sub-tree with the existing terminal node and the new leaf node forming the 
+children of the last branch node in the sub-tree. The depth of the sub-tree is determined by the number of matching 
+bits of the respective keys of the two nodes.
+
+To delete a node, the procedure above is reversed. This entails that a significant portion of the tree may be pruned 
+when deleting a node in a highly sparse region of the tree.
+
+The null hashes representing the empty sub-trees are treated identically to the leaf nodes. 
+Thus branch hashes are calculated in the usual way, _inter alia_, `H_branch = H(Branch marker, H_left, H_right)`, 
+irrespective of whether the left or right nodes are empty or not. 
+
+Domain separation SHOULD be used to distinguish branch nodes from leaf nodes. This also mitigates second pre-image 
+attacks if the advice above is not followed and the values are hashed with the same algorithm as the tree. 
+
+For leaf node hashes, the key MUST be included in the hash. This prevents leaf node spoofing of the pruned tree. 
+Therefore, leaf node hashes are of the form `H_leaf = H(Leaf marker, H_key, H_value)`.
+
+A proof of concept implementation has been written and submitted for review in 
+[PR #5457](https://github.com/tari-project/tari/pull/5457). The examples that follow assume this implementation.
+
+
+#### Example
+Let's create a SMT with four nodes.
+
+If we insert the nodes at
+ * A: 01001111 (79 in decimal)
+ * B: 01011111 (95 in decimal)
+ * C: 11100000 (224 in decimal)
+ * D: 11110000 (240 in decimal)
+
+you will notice that the first two diverge at the first bit, while the first and last pairs differ at the
+fourth bit. This results in a SMT that looks like this:
```

Comment:
The first and second indexes differ at the fourth bit, whereas the first and last pairs differ at the first bit.